### PR TITLE
fix: remove past hearings from CCD event (FPLA-1178)

### PR DESCRIPTION
**Pull request by swalker125**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/FPLA-1178

### Change description ###

Removes past hearings from hearingDetails collection, placing them into a hidden pastHearingDetails collection whilst the user completes the add hearing details event. This has no effect on the tab view of hearings.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by tomaszpowroznik** [comment]
Personally i would use caseDetailsBefore to find past hearings and remove them from hearing list in about to start, and then find same hearing in caseDetailsBefore in about to submit and add them back. Then new ccd field definition is not needed

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java [hunk 1]
```diff
@@ -132,29 +123,79 @@ private DynamicListElement createDynamicElement(UUID code) {
         }
     }
 
+    @Nested
+    class GetPastHearings {
+
+        @Test
+        void shouldReturnEmptyListWhenNoHearingsHaveBeenCreated() {
+            assertThat(service.getPastHearings(emptyList())).isEqualTo(emptyList());
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java]: minor, isEqualTo(emptyList()) can be replaced with .isEmpty()

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java [hunk 2]
```diff
@@ -132,29 +123,79 @@ private DynamicListElement createDynamicElement(UUID code) {
...
+        }
+
+        @Test
+        void shouldReturnEmptyListWhenNoPastHearingsExist() {
+            List<Element<HearingBooking>> hearingBooking = newArrayList(hearingElementWithStartDate(+5));
+
+            assertThat(service.getPastHearings(hearingBooking)).isEqualTo(emptyList());
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java]: same here isEmpty() should do the job

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java [hunk 3]
```diff
@@ -132,29 +123,79 @@ private DynamicListElement createDynamicElement(UUID code) {
...
-                .value(createHearingBooking(TODAYS_DATE, TODAYS_DATE.plusDays(1)))
-                .build()
+        return List.of(
+            element(LIST_UUID[0], createHearingBooking(TODAY.plusDays(5), TODAY.plusDays(6))),
+            element(LIST_UUID[1], createHearingBooking(TODAY.plusDays(2), TODAY.plusDays(3))),
+            element(LIST_UUID[2], createHearingBooking(TODAY, TODAY.plusDays(1)))
         );
     }
+
+    private Element<HearingBooking> hearingElementWithStartDate(int i) {
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java]: int daysFromToday instead on int i

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java [hunk 4]
```diff
@@ -132,29 +123,79 @@ private DynamicListElement createDynamicElement(UUID code) {
...
+            Element<HearingBooking> pastHearingBooking = hearingElementWithStartDate(-5);
+
+            List<Element<HearingBooking>> hearingBookings = newArrayList(futureHearingBooking, pastHearingBooking);
+
+            assertThat(service.getPastHearings(hearingBookings)).isEqualTo(List.of(pastHearingBooking));
+        }
+
+        private ArrayList<Element<HearingBooking>> emptyHearingBooking() {
+            return newArrayList(element(HearingBooking.builder().build()));
+        }
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java]: not used

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java [hunk 5]
```diff
@@ -132,29 +123,79 @@ private DynamicListElement createDynamicElement(UUID code) {
         }
     }
 
+    @Nested
+    class GetPastHearings {
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java]: maybe edge case where hearing date is today should be tested

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/HearingBookingKeys.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package uk.gov.hmcts.reform.fpl.enums;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+@Getter
+@AllArgsConstructor
+public enum HearingBookingKeys {
+    HEARING_DETAILS("hearingDetails");
```
Comment (tomaszpowroznik) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/enums/HearingBookingKeys.java]: minor, now enum does not make a lot of sense, maybe just string constant

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToStartTest.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        super("add-hearing-bookings");
+    }
+
+    @Test
+    void shouldReturnPopulatedHearingWhenNoOtherHearingsExist() {
+        AboutToStartOrSubmitCallbackResponse response = postAboutToStartEvent(callbackRequest(new HashMap<>()));
+
+        CaseData caseData = mapper.convertValue(response.getData(), CaseData.class);
+
+        assertThat(caseData.getHearingDetails()).hasSize(1);
```
Comment (tomaszpowroznik) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToStartTest.java]: this assertion is not strong enough, does not check that default hearing is empty, maybe:
assertThat(ElementUtils.unwrapElements(caseData.getHearingDetails())).containsExactly(HearingBooking.builder().build());

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToStartTest.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        List<Element<HearingBooking>> hearingDetails = newArrayList(
+            bookingWithStartDate(TODAY.plusDays(5)), bookingWithStartDate(TODAY.minusDays(5)));
+
+        AboutToStartOrSubmitCallbackResponse response =
+            postAboutToStartEvent(callbackRequest(Map.of(HEARING_DETAILS.getKey(), hearingDetails)));
+
+        CaseData caseData = mapper.convertValue(response.getData(), CaseData.class);
+
+        assertThat(caseData.getHearingDetails()).hasSize(1);
+        assertThat(caseData.getHearingDetails().get(0).getValue().getStartDate()).isAfter(TODAY);
```
Comment (tomaszpowroznik) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToStartTest.java]: you can have stronger and cleaner assertion with:

 Element<HearingBooking> futureHearingBooking = bookingWithStartDate(TODAY.plusDays(5));
        Element<HearingBooking> pastHearingBooking = bookingWithStartDate(TODAY.minusDays(5));
        
        List<Element<HearingBooking>> hearingDetails = newArrayList(futureHearingBooking, pastHearingBooking);

        AboutToStartOrSubmitCallbackResponse response =
            postAboutToStartEvent(callbackRequest(Map.of(HEARING_DETAILS.getKey(), hearingDetails)));

        CaseData caseData = mapper.convertValue(response.getData(), CaseData.class);

        assertThat(caseData.getHearingDetails()).containsExactly(futureHearingBooking);

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToStartTest.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    private CallbackRequest callbackRequest(Map<String, Object> data) {
+        return CallbackRequest.builder()
+            .caseDetails(CaseDetails.builder()
+                .data(data)
+                .build())
+            .build();
```
Comment (tomaszpowroznik) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToStartTest.java]: postAboutToStartEvent accepts CaseDetails so you dont have to wrap it here with CallbackRequest, couple lines of code will be saved

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToSubmitTest.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.element;
+
+@ActiveProfiles("integration-test")
+@WebMvcTest(HearingBookingDetailsController.class)
+@OverrideAutoConfiguration(enabled = true)
+class HearingBookingDetailsControllerAboutToSubmitTest extends AbstractControllerTest {
+    private static final LocalDateTime TODAY = LocalDateTime.now();
+
+    @Autowired
+    private ObjectMapper mapper;
```
Comment (tomaszpowroznik) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToSubmitTest.java]: not needed, parent class have it exposed

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToSubmitTest.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+    HearingBookingDetailsControllerAboutToSubmitTest() {
+        super("add-hearing-bookings");
+    }
+
+    @Test
+    void shouldReturnEmptyListWhenSubmittingEmptyEvent() {
```
Comment (tomaszpowroznik) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToSubmitTest.java]: EmptyHearingListWhenNoHearingInCase

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToSubmitTest.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    @Test
+    void shouldReturnHearingsWhenNoHearingsExistInPast() {
+        List<Element<HearingBooking>> hearingDetails = newArrayList(bookingWithStartDate(TODAY.plusDays(5)));
+
+        AboutToStartOrSubmitCallbackResponse response = postAboutToSubmitEvent(callbackRequest(
+            Map.of(HEARING_DETAILS.getKey(), hearingDetails), Map.of(HEARING_DETAILS.getKey(), hearingDetails)));
+
+        CaseData caseData = mapper.convertValue(response.getData(), CaseData.class);
+
+        assertThat(caseData.getHearingDetails()).hasSize(1);
```
Comment (tomaszpowroznik) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToSubmitTest.java]: not needed, 2nd assertion already verify it

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToSubmitTest.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    @Test
+    void shouldReturnHearingsWhenNoHearingsExistInFuture() {
+        List<Element<HearingBooking>> hearingDetails = newArrayList(bookingWithStartDate(TODAY.plusDays(-5)));
+
+        AboutToStartOrSubmitCallbackResponse response = postAboutToSubmitEvent(callbackRequest(
+            Map.of(HEARING_DETAILS.getKey(), emptyList()), Map.of(HEARING_DETAILS.getKey(), hearingDetails)));
+
+        CaseData caseData = mapper.convertValue(response.getData(), CaseData.class);
+
+        assertThat(caseData.getHearingDetails()).hasSize(1);
```
Comment (tomaszpowroznik) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToSubmitTest.java]: not needed, 2nd assertion verify it

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToSubmitTest.java [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+
+        AboutToStartOrSubmitCallbackResponse response = postAboutToSubmitEvent(callbackRequest(
+            Map.of(HEARING_DETAILS.getKey(), newArrayList(hearingDetail)),
+            Map.of(HEARING_DETAILS.getKey(), newArrayList(hearingDetail, hearingDetailPast))));
+
+        CaseData caseData = mapper.convertValue(response.getData(), CaseData.class);
+
+        assertThat(caseData.getHearingDetails()).hasSize(2);
+        assertThat(caseData.getHearingDetails().get(0)).isEqualTo(hearingDetailPast);
+        assertThat(caseData.getHearingDetails().get(1)).isEqualTo(hearingDetail);
```
Comment (tomaszpowroznik) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToSubmitTest.java]: assertThat(caseData.getHearingDetails()).containsExactly(hearingDetailPast, hearingDetail);

is enough, it assert order of elements

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/HearingBookingService.java [hunk 15]
```diff
@@ -6,41 +6,47 @@
...
+        return hearingDetails != null ? hearingDetails : EMPTY_HEARING_BOOKING;
+    }
+
+    public List<Element<HearingBooking>> getPastHearings(List<Element<HearingBooking>> hearingDetails) {
+        return hearingDetails.stream().filter(this::isBeforeToday).collect(toList());
+    }
+
+    private boolean isBeforeToday(Element<HearingBooking> element) {
+        return element.getValue() != null && element.getValue().getStartDate() != null
+            && element.getValue().getStartDate().isBefore(LocalDateTime.now());
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/HearingBookingService.java]: happy to update 👍  I have seen this Optionals pop up elsewhere

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/HearingBookingService.java [hunk 16]
```diff
@@ -6,41 +6,47 @@
...
 
+import static com.google.common.collect.Lists.newArrayList;
 import static java.util.Comparator.comparing;
+import static java.util.stream.Collectors.toList;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.element;
 
 @Service
 public class HearingBookingService {
+    private static final List<Element<HearingBooking>> EMPTY_HEARING_BOOKING =
+        newArrayList(element(HearingBooking.builder().build()));
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/HearingBookingService.java]: line 51 will throw a unsupported operation the first time we click through the event as we cannot call .remove on an immutable list, what's better here. To leave as is, or to wrap the remove in a if check?

[the remark this answers is not in the document]
Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/HearingBookingService.java]: just use mutable list but return fresh copy each time.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java [hunk 17]
```diff
@@ -118,12 +110,11 @@ void shouldReturnTheFirstHearingBookingWhenTheDynamicListValueCodeMatches() {
...
-                .value(createDynamicElement(UUIDS[0]))
+                .value(createDynamicElement(LIST_UUID[0]))
                 .listItems(List.of(
-                    createDynamicElement(UUIDS[0]),
-                    createDynamicElement(UUIDS[1]),
-                    createDynamicElement(UUIDS[2])
-                ))
+                    createDynamicElement(LIST_UUID[0]),
+                    createDynamicElement(LIST_UUID[1]),
+                    createDynamicElement(LIST_UUID[2])))
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java]: happy to update

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java (hunk 1 above)
Comment (swalker125) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java]: yep, silly of me

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java (hunk 5 above)
Comment (swalker125) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/HearingBookingServiceTest.java]: hmm, what should be expected here. Now will always be in the past by the time the assertion runs 🤔

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/HearingBookingKeys.java (hunk 6 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/enums/HearingBookingKeys.java]: Yeah enum is a bit overkill, you're right

Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/enums/HearingBookingKeys.java]: you ok for this to exist publicly in hearing booking service?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToStartTest.java (hunk 8 above)
Comment (swalker125) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToStartTest.java]: Already made a similar change in aboutToSubmit, shall do the same here 👍

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToStartTest.java [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.unwrapElements;
+
+@ActiveProfiles("integration-test")
+@WebMvcTest(HearingBookingDetailsController.class)
+@OverrideAutoConfiguration(enabled = true)
+class HearingBookingDetailsControllerAboutToStartTest extends AbstractControllerTest {
+    private static final LocalDateTime TODAY = LocalDateTime.now();
+
+    @Autowired
+    private ObjectMapper mapper;
```
Comment (tomaszpowroznik) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsControllerAboutToStartTest.java]: not needed

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsController.java [hunk 19]
```diff
@@ -58,13 +62,29 @@ public AboutToStartOrSubmitCallbackResponse handleMidEvent(@RequestBody Callback
...
+    }
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+        CaseDetails caseDetailsBefore = callbackRequest.getCaseDetailsBefore();
+        CaseData caseDataBefore = mapper.convertValue(caseDetailsBefore.getData(), CaseData.class);
+
+        List<Element<HearingBooking>> hearingDetailsBefore = service.expandHearingBookingCollection(caseDataBefore);
```
Comment (tomaszpowroznik) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/HearingBookingDetailsController.java]: is that ok? if user removes all future hearings, then default one will be added, this is not what we want here as empty hearing will be saved and displayed in tab?
